### PR TITLE
qcow-efi format with ESP mount and configurable diskSize

### PR DESCRIPTION
**Pull request by tylerjl**
Hey folks! This PR adds the qcow-efi format based on @ashraffouda's work and @Mic92's feedback in #194. Credit to both of them for the discussing and existing work in the PR, I just consolidated it into my branch since the work in #194 appeared stalled.

Summary of these changes:

- I started wtih the commit from @ashraffouda, which adds the baseline qcow-efi image
- I've been using their work for a while and found a bug that stopped `nixos-rebuild ... switch` from working due to the missing EFI mount point, so I added `/boot`, which fixes that.
- @Mic92 mentioned that exposing the disk image size as an option would be a good thing to do. That makes sense, so I wrapped the format in a traditional `options = { ... }; config = { ... };` form. The existing format modules do not do any of this so I had to make it up; I didn't choose to nest the `diskSize` option under any specific namespace/attribute but if there's any sort of preference it's all the same to me.

I'm actively using this format, so I know it works, but if there's anything additional y'all would like to see just let me know.

### CODE REVIEW

File: formats/qcow-efi.nix [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  config = {
+    fileSystems."/" = {
+      device = "/dev/disk/by-label/nixos";
+      autoResize = true;
+      fsType = "ext4";
+    };
+
+    fileSystems."/boot" = {
+      device = "/dev/disk/by-label/ESP";
```
Comment (Mic92) [inline formats/qcow-efi.nix:24]: I am wondering if this should just be a small efi ESP partition just for the grub efi executable without kernel and mounted to `/boot/efi` with `boot.loader.grub.efiSysMountPoint = "/boot/efi";`. The advantage is that the kernel than can reside on the ext4 partition which might have more space and in general is a better/safer filesystem.

Comment (Mic92) [reply formats/qcow-efi.nix:24]: cc @Lassulus

File: formats/qcow-efi.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    boot.kernelParams = [ "console=ttyS0" ];
+    boot.loader.grub.device = "nodev";
+
+    boot.loader.grub.efiSupport = true;
+    boot.loader.grub.efiInstallAsRemovable = true;
+    boot.loader.timeout = 0;
+
+    system.build.qcow = import "${toString modulesPath}/../lib/make-disk-image.nix" {
+      inherit lib config pkgs;
+      diskSize = config.diskSize;
```
Comment (Mic92) [inline formats/qcow-efi.nix:38]: If we go with the small boot size than the `bootSize` parameter could be reduced a bit.

File: formats/qcow-efi.nix [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      fsType = "ext4";
+    };
+
+    fileSystems."/boot" = {
+      device = "/dev/disk/by-label/ESP";
+      fsType = "vfat";
+    };
+
+    boot.growPartition = true;
+    boot.kernelParams = [ "console=ttyS0" ];
```
Comment (Mic92) [inline formats/qcow-efi.nix:29]: There are some other architecture specific console paths to consider: https://github.com/numtide/srvos/blob/6ee5c9e530453161d48b7c853ada2d9fefcf0017/common/serial.nix#L37

I think qemu on aarch64 for example uses: `ttyAMA0`

**Comment by RaitoBezarius** [comment]
Hey, can I do something to help you get this PR to the finish line? I'm interested into it as I'm running into limitations with the existing qcow biosboot only for x86_64-linux. :)

File: formats/qcow-efi.nix (hunk 3 above)
Comment (tylerjl) [reply formats/qcow-efi.nix:29]: Hm. Is this abstracted somewhere that we can reference, or should we copypasta the list from srvos?

Comment (Lassulus) [reply formats/qcow-efi.nix:29]: I think copy-paste is the most viable solution right now

File: formats/qcow-efi.nix (hunk 2 above)
Comment (tylerjl) [reply formats/qcow-efi.nix]: Addressed in [062cb2f](https://github.com/nix-community/nixos-generators/pull/204/commits/062cb2f55564ff53b3381d7a07369f469403df6e)

File: formats/qcow-efi.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{ config, lib, pkgs, modulesPath, ... }:
+{
+  # for virtio kernel drivers
+  imports = [
+    "${toString modulesPath}/profiles/qemu-guest.nix"
+  ];
+
+  options = {
+    diskSize = lib.mkOption {
```
[the remark this answers is not in the document]
Comment (tylerjl) [reply formats/qcow-efi.nix]: Addressed in [062cb2f](https://github.com/nix-community/nixos-generators/pull/204/commits/062cb2f55564ff53b3381d7a07369f469403df6e)

File: formats/qcow-efi.nix (hunk 3 above)
Comment (tylerjl) [reply formats/qcow-efi.nix]: Addressed in [1424032](https://github.com/nix-community/nixos-generators/pull/204/commits/142403228301f0af5f1709a8b1c6790f948d53ee)

**Comment by Lassulus** [comment]
somehow this fell under my radar, but this looks good to me.

**Comment by tylerjl** [comment]
Thank you so much, @Lassulus!

**Pull request merged**